classes/config_settings.php

```diff
@@ -0,0 +1,259 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Data object containing SEB config settings with defaults if required.
+ *
+ * @package    quizaccess_seb
+ * @author     Andrew Madden <andrewmadden@catalyst-au.net>
+ * @copyright  2019 Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace quizaccess_seb;
+
+defined('MOODLE_INTERNAL') || die();
+
+class config_settings {
+
+    /** @var string SETTINGS_TABLE DB table that contains quiz settings. */
+    const SETTINGS_TABLE = 'quizaccess_seb_quizsettings';
+
+    /** @var int $quizid Quiz ID of quiz the settings relate to. */
+    private $quizid;
+
+    /** @var \stdClass $settings Settings used to create class. */
+    private $settings;
+
+    /** @var int $existingsettingsid ID of existing settings in table. Defaults to -1 if none exist. */
+    private $existingsettingsid = -1;
+
+    /**
+     * config_settings constructor.
+     *
+     * @param int $quizid ID of quiz that the settings belong to.
+     * @param \stdClass|null $settings New quiz settings. May be provided by quiz settings form.
+     * @throws \dml_exception
+     * @throws \moodle_exception
+     */
+    public function __construct(int $quizid, \stdClass $settings = null) {
+        global $DB;
+
+        $this->quizid = $quizid;
+
+        // Try and retrieve existing settings, otherwise assign null.
+        $existingsettings = $DB->get_record(self::SETTINGS_TABLE, ['quizid' => $quizid], '*');
+        if (empty($existingsettings)) {
+            $existingsettings = new \stdClass();
+        } else {
+            $this->existingsettingsid = $existingsettings->id;
+        }
+
+        // If new settings are empty, set existing settings, otherwise merge existing settings with new settings.
+        if (empty((array) $settings)) {
+            $this->settings = $existingsettings;
```

Comment:
As settings could be potentially empty. Do we want to update DB definition and have default values of required fields?

Reply:
Ah I can see that you validate it later. But still doesn't check if it matches DB type.